src/Generator/Generators/CSharp/CSharpSources.cs

```diff
@@ -764,24 +767,63 @@ public override void GenerateClassSpecifier(Class @class)
                 Write(" : {0}", string.Join(", ", bases));
         }
 
-        private void GenerateClassInternalsField(LayoutField field, Class @class)
+        private bool CanUseSequentialLayout(Class @class)
         {
-            TypePrinterResult retType = TypePrinter.VisitFieldDecl(
-                new Field { Name = field.Name, QualifiedType = field.QualifiedType });
+            if (@class.IsUnion)
+                return false;
 
-            PushBlock(BlockKind.Field);
+            var fields = @class.Layout.Fields;
 
-            if (!Options.GenerateSequentialLayout || @class.IsUnion)
-                WriteLine($"[FieldOffset({field.Offset})]");
-            Write($"internal {retType}");
-            if (field.Expression != null)
-            {
-                var fieldValuePrinted = field.Expression.CSharpValue(ExpressionPrinter);
-                Write($" = {fieldValuePrinted}");
+            if (fields.Count > 1)
+            { 
+                for (var i = 1; i < fields.Count; ++i)
+                {
+                    if (fields[i].Offset == fields[i - 1].Offset)
+                        return false;
+                }
             }
-            WriteLine(";");
 
-            PopBlock(NewLineKind.BeforeNextBlock);
+            return true;
+        }
+
+        private void GenerateClassInternalsFields(Class @class, bool sequentalLayout)
+        {
+            var fields = @class.Layout.Fields;
+
+            for (var i = 0; i < fields.Count; ++i)
+            {
+                var field = fields[i];
+
+                TypePrinterResult retType = TypePrinter.VisitFieldDecl(
+                    new Field { Name = field.Name, QualifiedType = field.QualifiedType });
+
+                PushBlock(BlockKind.Field);
+
+                if (!sequentalLayout)
+                    WriteLine($"[FieldOffset({field.Offset})]");
+
+                Write($"internal {retType}");
+                if (field.Expression != null)
+                {
+                    var fieldValuePrinted = field.Expression.CSharpValue(ExpressionPrinter);
+                    Write($" = {fieldValuePrinted}");
+                }
+                WriteLine(";");
+
+                if (sequentalLayout && fields[i].IsVTablePtr)
+                {
+                    var nativePointerSize = (Context.TargetInfo.PointerWidth / 8);
+                    var padding = 
+                        i + 1 < fields.Count
+                        ? fields[i + 1].Offset - field.Offset - nativePointerSize
+                        : (uint)@class.Layout.Size - nativePointerSize;
+
+                    if (padding > 0 && padding <= @class.Layout.Size)
+                        WriteLine($"public fixed byte {field.Name}Padding[{padding}];");
```

Comment:
I didn't get why this padding is necessary. It looks like a bug, a field missing from the AST.